src/kpro_lib.erl

```diff
@@ -0,0 +1,301 @@
+get_ts_type(_, A) when ?KPRO_IS_APPEND_TS(A) -> append.
+
+%% os:system_time(millisecond) is since otp 19
+-spec now_ts() -> kpro:msg_ts().
+now_ts() -> os:system_time() div 1000000.
+
+-spec get_req_schema(kpro:api(), kpro:vsn()) -> kpro:struct_schema().
+get_req_schema(Api, Vsn) ->
+  F = fun() -> kpro_schema:req(Api, Vsn) end,
+  get_schema(F, {Api, req, Vsn}).
+
+-spec get_rsp_schema(kpro:api(), kpro:vsn()) -> kpro:struct_schema().
+get_rsp_schema(Api, Vsn) ->
+  F = fun() -> kpro_schema:rsp(Api, Vsn) end,
+  get_schema(F, {Api, rsp, Vsn}).
+
+-spec get_prelude_schema(atom(), kpro:vsn()) -> kpro:struct_schema().
+get_prelude_schema(Tag, Vsn) ->
+  F = fun() -> kpro_prelude_schema:get(Tag, Vsn) end,
+  get_schema(F, {Tag, Vsn}).
+
+-spec find(kpro:field_name(), kpro:struct(), term()) -> kpro:field_value().
+find(Field, Struct, Error) when is_map(Struct) ->
+  try
+    maps:get(Field, Struct)
+  catch
+    error : {badkey, _} ->
+      erlang:error(Error)
+  end;
+find(Field, Struct, Error) when is_list(Struct) ->
+  case lists:keyfind(Field, 1, Struct) of
+    {_, Value} -> Value;
+    false -> erlang:error(Error)
+  end;
+find(_Field, Other, _Error) ->
+  erlang:error({not_struct, Other}).
+
+%% @doc Equivalent to `maps:update_with/4' (since otp 19).
+update_map(Key, Fun, Init, Map) ->
+  case Map of
+    #{Key := Value} -> Map#{Key := Fun(Value)};
+    _ -> Map#{Key => Init}
+  end.
+
+%% @doc delegate function evaluation to a agent process
+%% abort if it does not finish in time.
+%% exceptions and linked processes are caught in agent process
+%% adn propagated to parent process
+with_timeout(F0, Timeout) ->
+  Parent = self(),
+  ResultRef = make_ref(),
+  AgentFun =
+    fun() ->
+        {links, Links0} = process_info(self(), links),
+        Result =
+          try
+            {normal, F0()}
+          catch
+            C : E ->
+              CrashContext = {C, E, erlang:get_stacktrace()},
+              {exception, CrashContext}
+          end,
+        {links, Links1} = process_info(self(), links),
+        Links = Links1 -- Links0,
+        Parent ! {ResultRef, Result, Links},
+        receive
+          done ->
+            %% parent is done linking to links
+            %% safe to unlink and exit
+            [unlink(Pid) || Pid <- Links],
+            exit(normal)
+        end
+      end,
+  Agent = erlang:spawn_link(AgentFun),
+  receive
+    {ResultRef, Result, Links} ->
+      %% replicate links from agent
+      %% TODO handle link/1 exception if any of the links are dead already
+      [link(Pid) || Pid <- Links],
+      unlink(Agent),
+      Agent ! done,
+      case Result of
+        {normal, Return} ->
+          Return;
+        {exception, {C, E, Stacktrace}} ->
+          %% replicate exception from agent
+          erlang:raise(C, E, Stacktrace)
+      end
+  after
+    Timeout ->
+      %% kill agent
+      unlink(Agent),
+      erlang:exit(Agent, kill),
+      {error, timeout}
```

Comment:
So the timeout case returns a plain tuple, while Agent errors cause an exception, and Agent success returns an untagged value.  This seems a bit inconsistent.  And F0() could in principle also return a 2-tuple, which could be confused with the timeout return tuple.

Reply:
This delegation code is not made generic for use cases beside `kpro_lib:ok_pipe`.
The 'contract' for `ok_pipe` functions is: they return `ok | {ok, Value} | {error, Reason}`.